File: client/java/Frame.java
import com.sun.jna.Pointer;

///A baby robot.
class Frame
{
  Pointer ptr;
  int ID;
  int iteration;
  public Frame(Pointer p)
  {
    ptr = p;
    ID = Client.INSTANCE.frameGetId(ptr);
    iteration = BaseAI.iteration;
  }

  boolean validify()
  {
    if(iteration == BaseAI.iteration) return true;
    for(int i = 0; i < BaseAI.frames.length; i++)
    {
      if(BaseAI.frames[i].ID == ID)
      {
        ptr = BaseAI.frames[i].ptr;
        iteration = BaseAI.iteration;
        return true;
      }
    }
    throw new ExistentialError();
  }
    
    //commands
    
  ///Sends a message to be visualized when this unit is selected
  int talk(String message)
  {
    validify();
    return Client.INSTANCE.frameTalk(ptr, message);
  }
    
    //getters
    
  ///Unique Identifier
  public int getId()
  {
    validify();
    return Client.INSTANCE.frameGetId(ptr);
  }
  ///The X position of the top left corner of this object.  X is horizontal
  public int getX()
  {
    validify();
    return Client.INSTANCE.frameGetX(ptr);
  }
  ///The Y position of the top left corner of this object.  Y is vertical
  public int getY()
  {
    validify();
    return Client.INSTANCE.frameGetY(ptr);
  }
  ///The owning player
  public int getOwner()
  {
    validify();
    return Client.INSTANCE.frameGetOwner(ptr);
  }
  ///How much health this unit currently has
  public int getHealth()
  {
    validify();
    return Client.INSTANCE.frameGetHealth(ptr);
  }
  ///The maximum amount of health this unit can ever have
  public int getMaxHealth()
  {
    validify();
    return Client.INSTANCE.frameGetMaxHealth(ptr);
  }
  ///What type this robot will be
  public Pointer getType()
  {
    validify();
    return Client.INSTANCE.frameGetType(ptr);
  }
  ///The length of one side of this robot, such that size^2 = number of bots combined into this bot
  public int getSize()
  {
    validify();
    return Client.INSTANCE.frameGetSize(ptr);
  }
  ///How many of your turns until this frame becomes a robot
  public int getCompletionTime()
  {
    validify();
    return Client.INSTANCE.frameGetCompletionTime(ptr);
  }

}

File: client/c/Frame.cpp
// -*-c++-*-

#include "Frame.h"
#include "game.h"

#include "Type.h"

Frame::Frame(_Frame* pointer)
{
    ptr = (void*) pointer;
}

int Frame::id()
{
  return ((_Frame*)ptr)->id;
}

int Frame::x()
{
  return ((_Frame*)ptr)->x;
}

int Frame::y()
{
  return ((_Frame*)ptr)->y;
}

int Frame::owner()
{
  return ((_Frame*)ptr)->owner;
}

int Frame::health()
{
  return ((_Frame*)ptr)->health;
}

int Frame::maxHealth()
{
  return ((_Frame*)ptr)->maxHealth;
}

_Type* Frame::type()
{
  return ((_Frame*)ptr)->type;
}

int Frame::size()
{
  return ((_Frame*)ptr)->size;
}

int Frame::completionTime()
{
  return ((_Frame*)ptr)->completionTime;
}


int Frame::talk(std::string message)
{
  return frameTalk( (_Frame*)ptr, message.c_str());
}



std::ostream& operator<<(std::ostream& stream,Frame ob)
{
  stream << "id: " << ((_Frame*)ob.ptr)->id  <<'\n';
  stream << "x: " << ((_Frame*)ob.ptr)->x  <<'\n';
  stream << "y: " << ((_Frame*)ob.ptr)->y  <<'\n';
  stream << "owner: " << ((_Frame*)ob.ptr)->owner  <<'\n';
  stream << "health: " << ((_Frame*)ob.ptr)->health  <<'\n';
  stream << "size: " << ((_Frame*)ob.ptr)->size  <<'\n';
  stream << "maxHealth: " << ((_Frame*)ob.ptr)->maxHealth  <<'\n';
  stream << "type: " << ((_Frame*)ob.ptr)->type  <<'\n';
  stream << "completionTime: " << ((_Frame*)ob.ptr)->completionTime  <<'\n';
  return stream;
}
